go/ql/lib/semmle/go/dataflow/internal/TaintTrackingUtil.qll

```diff
@@ -27,11 +27,21 @@ predicate localExprTaint(Expr src, Expr sink) {
  * Holds if taint can flow in one local step from `src` to `sink`.
  */
 predicate localTaintStep(DataFlow::Node src, DataFlow::Node sink) {
-  DataFlow::localFlowStep(src, sink) or
-  localAdditionalTaintStep(src, sink, _) or
+  DataFlow::localFlowStep(src, sink)
+  or
+  localAdditionalTaintStep(src, sink, _)
+  or
   // Simple flow through library code is included in the exposed local
   // step relation, even though flow is technically inter-procedural
   FlowSummaryImpl::Private::Steps::summaryThroughStepTaint(src, sink, _)
+  or
+  // Treat container flow as taint for the local taint flow relation
+  exists(DataFlow::Content c | DataFlowPrivate::containerContent(c) |
```

Comment:
Why is this defined here and not in localAdditionalTaintStep?

Reply:
In increasing order of how good an answer it is:
(a) It's what they do in java.
(b) I copied it before I realised that it wouldn't affect global taint flow at all.
(c) I think it is helpful. Six months ago I tried to convert old-style models for variadic functions to MaD. [One of the sticking points](https://github.com/github/codeql/pull/16592#issuecomment-2136000245) was "Some barrier guard definitions use localTaintStep to detect taint flow, but it doesn't work flow through an implicit varargs slice." With this added (for container reads and container writes) we get that `localTaint` includes flow into and out of arrays, including through implicit varargs slices.

Reply:
OK, I think I see: you want both reads and stores to appear according to the handy utility `localTaintStep`, but only reads to be contributed to the additional-taint-step for global dataflow's purposes?

If so, suggest two things:
1. I note @aschackmull suggested the summary-getter step should be passed to the dataflow lib (i.e. in isAdditionalTaintStep`)
2. You might as well omit adding the read steps again here since `localAdditionalTaintStep` is already included -- instead just comment that the read steps are already taken care of
3. Add a comment explaining why we want read steps given to the dataflow lib but not stores

Reply:
1. The QLDoc for `summaryGetterStep` says

>        * NOTE: This step should not be used in global data-flow/taint-tracking, but may
>        * be useful to include in the exposed local data-flow/taint-tracking relations.

2. Done.
3. I'm not exactly sure what to say. I've put "Treat container read steps as taint for global taint flow." @aschackmull can you suggest a more detailed explanation?

Reply:
As the qldoc that Owen referenced says, the summary-getter step should _not_ be passed to global data flow in any form (neither as a read step nor as a taint step). To elaborate: Container read steps should be passed to data flow as taint steps. But read steps and summary-getter steps are separate things: A read step is atomic, whereas a summary-getter step represents a flow path through a callable with a nested read step, such that the entire thing can be summarized and treated as a read step in those contexts that aren't capable of doing such summarization. And data flow notably is perfectly capable of such summarization, so it only needs to know about the atomic read steps.